tests/test_templates_broadcast.py

```diff
@@ -0,0 +1,267 @@
+
+    @pytest.mark.parametrize("structure, unitary, parameters", GATE_PARAMETERS)
+    def test_correct_queue_for_gate_unitary(self, structure, unitary, parameters):
+        """Tests that correct gate queue is created when 'block' is a single gate."""
+
+        with qml.utils.OperationRecorder() as rec:
+            broadcast(block=unitary, structure=structure, wires=range(3), parameters=parameters)
+
+        for gate in rec.queue:
+            assert isinstance(gate, unitary)
+
+    @pytest.mark.parametrize("structure, unitary, gates, parameters",
+                             [("single", ParametrizedTemplate, [RX, RY], [[0.1, 1], [0.2, 1], [0.1, 1]]),
+                              ("single", ConstantTemplate, [T, S], [[], [], []]),
+                              ("double", ParametrizedTemplateDouble, [CRX, RY], [[0.1, 1]]),
+                              ("double", ConstantTemplateDouble, [T, CNOT], [[]]),
+                              ])
+    def test_correct_queue_for_template_block(self, structure, unitary, gates, parameters):
+        """Tests that correct gate queue is created when 'block' is a template."""
+
+        with qml.utils.OperationRecorder() as rec:
+            broadcast(block=unitary, structure=structure, wires=range(3), parameters=parameters)
+
+        first_gate = gates[0]
+        second_gate = gates[1]
+        for idx, gate in enumerate(rec.queue):
+            if idx % 2 == 0:
+                assert isinstance(gate, first_gate)
+            else:
+                assert isinstance(gate, second_gate)
+
+    @pytest.mark.parametrize("structure, template, kwarg, target_queue, parameters",
+                             [("single", KwargTemplate, True, [T, RY, T, RY], [[1], [2]]),
+                              ("single", KwargTemplate, False, [RY, RY], [[1], [2]]),
+                              ("double", KwargTemplateDouble, True, [T, CRX], [[1]]),
+                              ("double", KwargTemplateDouble, False, [CRX], [[1]])])
+    def test_correct_queue_for_template_block_with_keyword(self, structure, template, kwarg, target_queue, parameters):
+        """Tests that correct gate queue is created when 'block' is a template that uses a keyword."""
+
+        with qml.utils.OperationRecorder() as rec:
+            broadcast(block=template, structure=structure, wires=range(2),
+                      parameters=parameters, kwargs={'a': kwarg})
+
+        for gate, target_gate in zip(rec.queue, target_queue):
+            assert isinstance(gate, target_gate)
+
+    @pytest.mark.parametrize("structure, pars1, pars2, gate", [("single", [[], [], []], None, T),
+                                                               ("single", [1, 2, 3], [[1], [2], [3]], RX),
+                                                               ])
+    def test_correct_queue_for_gate_block(self, structure, pars1, pars2, gate):
+        """Tests that specific parameter inputs have the same output."""
```

Comment:
So just to make sure: this is testing that the same gate can be broadcast by passing arguments to `parameters` that are merely different in structure (e.g. list vs list of lists, list of empty lists vs `None`)?

Reply:
If so, maybe the name could be something like `test_correct_queue_same_gate_block_different_parameter_formats`, as you were using the `_different_parameter_formats` naming later as well :slightly_smiling_face:

Reply:
Ah thanks for pointing that out. Having this test for both structures is actually redundant. I renamed as you said and deleted the test for `double`.

Reply:
Oh, actually, in the sense that it tested that the specific block was a `single` block, it was perhaps not entirely redundant. Though hopefully might also not be the most important of tests. :)

Reply:
Also just saw your comment on the integration test cases, that would be another reason to leave it as is